README
Porting of SuperTux 0.1.3 to Miyoo (Bittboy / PocketGo / PowKiddy V90/Q90)

Controls (PocketGo):
- D-Pad: Move
- B: Jump/Confirm
- Y: Run/Action
- Select: Menu

Building from source:
  make MIYOO=1

If you have not received a copy of the source code and license agreement with this binary release,
you can visit https://github.com/andrigamerita/supertux


--- --- ---


- An introduction for SuperTux -
http://super-tux.sf.net/

Last update: April 26, 2004
 
DESCRIPTION
-----------
  SuperTux is a jump'n run like game, with strong inspiration from the
  Super Mario Bros games for Nintendo.

  Run and jump through multiple worlds, fighting off enemies by jumping
  on them or bumping them from below.  Grabbing power-ups and other stuff
  on the way.


STORY
-----

- Penny gets captured! -

  Tux and Penny were out having a nice
  picnic on the ice fields of Antarctica.
  Suddenly, a creature jumped from
  behind an ice bush, there was a flash,
  and Tux fell asleep!

  When Tux wakes up, he finds that Penny
  is missing. Where she lay before now
  lies a letter. "Tux, my arch enemy!"
  says the letter. "I have captured
  your beautiful Penny and have taken her
  to my fortress. The path to my fortress
  is littered with my minions. Give up on
  the thought of trying to reclaim her,
  you haven't a chance! -Nolok"

  Tux looks and see Nolok's fortress in
  the distance. Determined to save his
  beloved Penny, he begins his journey.


DOCUMENTATION
-------------
  Important documentation for SuperTux is contained in multiple files.
  Please see them:

    AUTHORS     - Credits for people that contributed to the creation
                  of SuperTux.
    NEWS        - Changes since the previous versions of "SuperTux"
    COPYING     - The GNU Public License, which "SuperTux" is liscensed
                  under.
    INSTALL     - Instructions on requirements, compiling and installing.
    README      - (This file.)  Game story, usage, rules and controls.
    TODO        - A wish-list for this version of "SuperTux"
    LEVELDESIGN - Very useful information for those that want to design
                  levels for SuperTux.


RUNNING THE GAME
----------------
  SuperTux makes use of proc to see where it is. In other words, it does not
  have any need to be installed and can be run from anywhere. This makes
  it possible to click in the executable in your filemanager (ie. Konqueror,
  Nautilus) as oposited to most of Linux games.

  Options can be reached from the menu, so you don't need to specify arguments,
  but if you want, type 'supertux --help' to check the ones that are avaible.
  Also, notice that SuperTux saves the options, so you don't need to be allways
  specifying them.

  The game has two frontends: SDL and OpenGL. Currently, there isn't any
  noticiable difference between them. Basically, SDL uses software to draw
  the game and OpenGL makes use of the videocards CPU (if the videocard has
  one and the drivers support it). Under linux, nVidia is recommendaded
  (with the closed drivers), but an ATI video card should work just fine.


PLAYING THE GAME
----------------
  Controls
  --------
  Both keyboards and joysticks/gamepads are supported. You can change the
  controls via the Options menu. Basically, the only keys you will need to
  use in-game are to do the following actions: jump, duck, right, left, power
  and 'P' to pause/unpause the game. I don't think there is anything to
  explain about the others, but the power key, also called run, allow you
  to run and use the power of the powerup you got. For instance, with the
  fire flower, you can shoot bullets (note that this is the only power
  currently implemented).
  
  Other useful keys include the Esc key, which is used to go to the menu or
  to go up a level in the menu. The menu can be navigated using the arrow
  keys or the mouse.

  In the worldmap, the arrow keys are used to navigate and Enter enters the
  current level.

  The leveleditor is controled via the mouse, but there are shortcuts that
  may make your life easier. To see the shortcuts for the buttons, just
  do a Right-click on them. If you are interested in creating SuperTux
  levels, we advise you to read LEVELDESIGN to learn more about the editor
  and creating new levels.


SCREEN LAYOUT
-------------
  Status
  ------
    The amount of time you have left to complete this level is displayed in
    the center at the top of the screen.  (Note: Time is NOT in seconds!)

    The number of coins you have collected is displayed at the upper right.

SCORING
-------
  Enemies
  -------
    If you kill more than one enemy at a time without landing on the ground,
    each enemy is worth more points.  For example, if there are three in a
    row and you jump on the first, it's worth 50 points.  If you bounce off
    the first and onto the second, it's worth 100 points.  If you also get
    the third, it's worth 150 points.

    You can also score by shooting at the enemies. Don't forget to get
    the iceflower in order to get the ability to fire.

THE END
-------
  Thanks for trying out SuperTux.

  If you like SuperTux, please rate and review it at the
  Linux Game Tome:

    http://www.happypenguin.org/

  Search for "SuperTux"!


  Thanks!

  The SuperTux development team


ChangeLog
- SuperTux Miyoo porting changes -
http://super-tux.sf.net/

Last update: August 27, 2021

0.2
- Fixed miyoo music loading issue, and made music loading fail safely

0.1
- Initial release.


src/sound.cpp
//  $Id$
//
//  SuperTux -  A Jump'n Run
//  Copyright (C) 2000 Bill Kendrick <bill@newbreedsoftware.com>
//  Copyright (C) 2004 Duong-Khang NGUYEN <neoneurone@users.sf.net>
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#include "defines.h"
#include "globals.h"
#include "sound.h"
#include "setup.h"
#ifdef USEMIKMOD
#include <string.h>
#endif

/*global variable*/
bool use_sound = true;    /* handle sound on/off menu and command-line option */
bool use_music = true;    /* handle music on/off menu and command-line option */
bool audio_device = true; /* != 0: available and initialized */

const char * soundfilenames[NUM_SOUNDS] = {
                                       "/sounds/jump.ogg",
                                       "/sounds/bigjump.ogg",
                                       "/sounds/skid.ogg",
                                       "/sounds/distro.ogg",
                                       "/sounds/herring.ogg",
                                       "/sounds/brick.ogg",
                                       "/sounds/hurt.ogg",
                                       "/sounds/squish.ogg",
                                       "/sounds/fall.ogg",
                                       "/sounds/ricochet.ogg",
                                       "/sounds/bump-upgrade.ogg",
                                       "/sounds/upgrade.ogg",
                                       "/sounds/excellent.ogg",
                                       "/sounds/coffee.ogg",
                                       "/sounds/shoot.ogg",
                                       "/sounds/lifeup.ogg",
                                       "/sounds/stomp.ogg",
                                       "/sounds/kick.ogg",
                                       "/sounds/explode.ogg",
                                       "/sounds/warp.ogg"
                                    };


#include <SDL_mixer.h>

#ifndef USEMIKMOD
Mix_Chunk * sounds[NUM_SOUNDS];
#else

#include <mikmod.h>
static MODULE *music=NULL;
static SAMPLE *chunk[NUM_SOUNDS];
static int chunkFlag[NUM_SOUNDS];
#endif

/* --- OPEN THE AUDIO DEVICE --- */

int open_audio (int frequency, Uint16 format, int channels, int chunksize)
{
//  close_audio();
#ifndef USEMIKMOD
  if (Mix_OpenAudio( frequency, format, channels, chunksize ) < 0)
    return -1;

  // allocate 16 channels for mixing
  if (Mix_AllocateChannels(8)  != 8)
    return -2;
  
  /* reserve some channels and register panning effects */
  if (Mix_ReserveChannels(SOUND_RESERVED_CHANNELS) != SOUND_RESERVED_CHANNELS)
    return -3;

  /* prepare the spanning effects */
  Mix_SetPanning( SOUND_LEFT_SPEAKER, 230, 24 );
  Mix_SetPanning( SOUND_RIGHT_SPEAKER, 24, 230 );
#else
  if (drv_oss.Name)  // Valid OSS driver
  {
    if (drv_oss.CommandLine)  // Valid Commandline
    {
      drv_oss.CommandLine("buffer=14,count=2");
    }
    MikMod_RegisterDriver(&drv_oss);
  }
  if (drv_alsa.Name)  // Valid ALSA driver
  {
    if (drv_alsa.CommandLine)  // Valid Commandline
    {
      drv_alsa.CommandLine("buffer=14");
    }
    MikMod_RegisterDriver(&drv_alsa);
  }
  MikMod_RegisterDriver(&drv_nos);

  // register standard tracker
  MikMod_RegisterAllLoaders();

  // Note, the md_mode flags are already set by default
  md_mode |= DMODE_SOFT_SNDFX | DMODE_SOFT_MUSIC;

   if (MikMod_Init(""))  // Command paramenters are ignored as all drivers are registered
  {
    printf("mikmod init war fürn arsch\n");
    return 1;
  }

  load_sounds();

  // get ready to play
  MikMod_EnableOutput();
  md_volume =64;
  md_sndfxvolume = 64;
  //md_musicvolume = md_sndfxvolume = 64;

  use_sound = true;
  audio_device=true;

#endif
  return 0;

}

/* --- CLOSE THE AUDIO DEVICE --- */

void close_audio( void )
{
#ifndef USEMIKMOD
  if (audio_device) {
    Mix_UnregisterAllEffects( SOUND_LEFT_SPEAKER );
    Mix_UnregisterAllEffects( SOUND_RIGHT_SPEAKER );
    Mix_CloseAudio();
  }
#else
  int i;
  if (! audio_device) return;

  MikMod_DisableOutput();
  Player_Stop();
  MikMod_Update();

  if (music) {
	  Player_Free(music);
	  music = NULL;
  }

  for ( i=0 ; i<NUM_SOUNDS ; i++ ) {
    if ( chunk[i] ) {
      Sample_Free(chunk[i]);
	}
  }

  MikMod_Update();
  MikMod_Exit();
#endif
}
#ifdef USEMIKMOD
void updateSound ( void ) {
  if (! audio_device) return;
  MikMod_Update();
}
#endif


/* --- LOAD A SOUND --- */

Mix_Chunk* load_sound(const std::string& file)
{
  if(!audio_device)
    return 0;
  
  Mix_Chunk* snd = Mix_LoadWAV(file.c_str());

  if (snd == 0)
    st_abort("Can't load", file);

  return(snd);
}
#ifdef USEMIKMOD
static void load_sounds() {
  int i;
  std::string name;
  char file[100];

  for ( i=0 ; i<NUM_SOUNDS ; i++ ) {
    name = datadir + soundfilenames[i];
    printf("loading: %s\n",name.c_str());
    snprintf(file,sizeof(file),"%s",name.c_str());
    if ( NULL == (chunk[i] = Sample_Load(file)) ) {
      fprintf(stderr, "Couldn't load: %s\n", file);
      use_sound = 0;
      return;
    }
    chunkFlag[i] = 0;
  }

  // reserve voices for sound effects
  MikMod_SetNumVoices(-1, 4);
}
#endif


/* --- PLAY A SOUND ON LEFT OR RIGHT OR CENTER SPEAKER --- */

void play_sound(Mix_Chunk * snd, enum Sound_Speaker whichSpeaker)
{
  /* this won't call the function if the user has disabled sound
   * either via menu or via command-line option
   */
  if(!audio_device || !use_sound)
    return;

  Mix_PlayChannel( whichSpeaker, snd, 0);

  /* prepare for panning effects for next call */
  switch (whichSpeaker) {
    case SOUND_LEFT_SPEAKER:
      Mix_SetPanning( SOUND_LEFT_SPEAKER, 230, 24 );
      break;
    case SOUND_RIGHT_SPEAKER:
      Mix_SetPanning( SOUND_RIGHT_SPEAKER, 24, 230 );
      break;
    default:  // keep the compiler happy
      break;
  }
}

#ifdef USEMIKMOD
void play_chunk(int idx) 
{
    int cid;
//  if (use_sound) return;
     cid = Sample_Play (chunk[idx], 0, 0);
  Voice_SetPanning(cid, PAN_CENTER);
  Voice_SetVolume(cid, 1000);
}
#endif


void free_chunk(Mix_Chunk *chunk)
{
  Mix_FreeChunk( chunk );
}

void sound_volume ( int vol )
{
#ifndef USEMIKMOD
    static int volume = 10;
    
    if ( vol == 1 ) volume-=5;
    else if ( vol == 2 ) volume+=5;
    Mix_Volume(-1,volume);
#endif
}
#ifdef USEMIKMOD
void increaseSoundVolume(void)
{
	if (md_volume > (256 - 13)) md_volume =  256;
	else md_volume += 13;
}

void decreaseSoundVolume(void)
{
	if (md_volume < 13) md_volume = 0;
	else md_volume -= 13;
}
#endif


src/player.cpp
//  $Id$
//
//  SuperTux -  A Jump'n Run
//  Copyright (C) 2003 Tobias Glaesser <tobi.web@gmx.de>
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#include <math.h>
#include "gameloop.h"
#include "globals.h"
#include "player.h"
#include "defines.h"
#include "scene.h"
#include "tile.h"
#include "sprite.h"
#include "screen.h"

#define AUTOSCROLL_DEAD_INTERVAL 300

Surface* tux_life;

Sprite* smalltux_gameover;
Sprite* smalltux_star;
Sprite* largetux_star;

PlayerSprite smalltux;
PlayerSprite largetux;
PlayerSprite firetux;

PlayerKeymap keymap;

PlayerKeymap::PlayerKeymap()
{
  keymap.jump  = SDLK_LCTRL;
  keymap.duck  = SDLK_DOWN;
  keymap.left  = SDLK_LEFT;
  keymap.right = SDLK_RIGHT;
  keymap.fire  = SDLK_SPACE; //SDLK_LALT;
}

void player_input_init(player_input_type* pplayer_input)
{
  pplayer_input->down = UP;
  pplayer_input->fire = UP;
  pplayer_input->left = UP;
  pplayer_input->old_fire = UP;
  pplayer_input->right = UP;
  pplayer_input->up = UP;
  pplayer_input->old_up = UP;
}

void
Player::init()
{
  Level* plevel = World::current()->get_level();

  holding_something = false;

  base.width = 32;
  base.height = 32;

  size = SMALL;
  got_coffee = false;

  base.x = plevel->start_pos_x;
  base.y = plevel->start_pos_y;
  base.xm = 0;
  base.ym = 0;
  previous_base = old_base = base;
  dir = RIGHT;
  old_dir = dir;
  duck = false;

  dying   = DYING_NOT;
  jumping = false;
  can_jump = true;

  frame_main = 0;
  frame_ = 0;
  
  player_input_init(&input);

  invincible_timer.init(true);
  skidding_timer.init(true);
  safe_timer.init(true);
  frame_timer.init(true);
  kick_timer.init(true);

  physic.reset();
}

int
Player::key_event(SDLKey key, int state)
{
  if(key == keymap.right)
    {
      input.right = state;
      return true;
    }
  else if(key == keymap.left)
    {
      input.left = state;
      return true;
    }
  else if(key == keymap.jump)
    {
      input.up = state;
      return true;
    }
  else if(key == keymap.duck)
    {
      input.down = state;
      return true;
    }
  else if(key == keymap.fire)
    {
      if (state == UP)
        input.old_fire = UP;
      input.fire = state;
      return true;
    }
  else
    return false;
}

void
Player::level_begin()
{
  base.x  = 100;
  base.y  = 170;
  base.xm = 0;
  base.ym = 0;
  previous_base = old_base = base;
  duck = false;

  dying = DYING_NOT;

  player_input_init(&input);

  invincible_timer.init(true);
  skidding_timer.init(true);
  safe_timer.init(true);
  frame_timer.init(true);

  physic.reset();
}

void
Player::action(double frame_ratio)
{
  bool jumped_in_solid = false;

  if (input.fire == UP)
    holding_something = false;

  /* Move tux: */
  previous_base = base;

  /* --- HANDLE TUX! --- */
  if(dying == DYING_NOT)
    handle_input();

  physic.apply(frame_ratio, base.x, base.y);

  if(dying == DYING_NOT) 
    {
      base_type target = base;

      collision_swept_object_map(&old_base, &base);

      // Don't accelerate Tux if he is running against a wall
      if (target.x != base.x)
        {
          physic.set_velocity_x(0);
        }

      // special exception for cases where we're stuck under tiles after
      // being ducked. In this case we drift out
      if(!duck && on_ground() && old_base.x == base.x && old_base.y == base.y
         && collision_object_map(base))
        {
          base.x += frame_ratio * WALK_SPEED * (dir ? 1 : -1);
          previous_base = old_base = base;
        }

      // Land:
      if (!on_ground())
        {
          physic.enable_gravity(true);
          if(under_solid())
            {
              // fall down
              physic.set_velocity_y(0);
              jumped_in_solid = true;
            }
        }
      else
        {
          /* Land: */
          if (physic.get_velocity_y() < 0)
            {
              base.y = (int)(((int)base.y / 32) * 32);
              physic.set_velocity_y(0);
            }

          physic.enable_gravity(false);
          /* Reset score multiplier (for multi-hits): */
          player_status.score_multiplier = 1;
        }

      if(jumped_in_solid)
        {
          if (isbrick(base.x, base.y) ||
              isfullbox(base.x, base.y))
            {
              World::current()->trygrabdistro(base.x, base.y - 32,BOUNCE);
              World::current()->trybumpbadguy(base.x, base.y - 64);

              World::current()->trybreakbrick(base.x, base.y, size == SMALL, RIGHT);

              bumpbrick(base.x, base.y);
              World::current()->tryemptybox(base.x, base.y, RIGHT);
            }

          if (isbrick(base.x+ 31, base.y) ||
              isfullbox(base.x+ 31, base.y))
            {
              World::current()->trygrabdistro(base.x+ 31, base.y - 32,BOUNCE);
              World::current()->trybumpbadguy(base.x+ 31, base.y - 64);

              if(size == BIG)
                World::current()->trybreakbrick(base.x+ 31, base.y, size == SMALL, LEFT);

              bumpbrick(base.x+ 31, base.y);
              World::current()->tryemptybox(base.x+ 31, base.y, LEFT);
            }
        }

      grabdistros();

      if (jumped_in_solid)
        {
          ++base.y;
          ++old_base.y;
          if(on_ground())
            {
              /* Make sure jumping is off. */
              jumping = false;
            }
        }
    }

//(USEMIKMOD patch)	updateSound();

  /* ---- DONE HANDLING TUX! --- */

  // check some timers
  skidding_timer.check();
  invincible_timer.check();
  safe_timer.check();
  kick_timer.check();
}

bool
Player::on_ground()
{
  return ( issolid(base.x + base.width / 2, base.y + base.height) ||
           issolid(base.x + 1, base.y + base.height) ||
           issolid(base.x + base.width - 1, base.y + base.height)  );
}

bool
Player::under_solid()
{
  return ( issolid(base.x + base.width / 2, base.y) ||
           issolid(base.x + 1, base.y) ||
           issolid(base.x + base.width - 1, base.y)  );
}

void
Player::handle_horizontal_input()
{
  float vx = physic.get_velocity_x();
  float vy = physic.get_velocity_y();
  float ax = physic.get_acceleration_x();
  float ay = physic.get_acceleration_y();

  float dirsign = 0;
  if(input.left == DOWN && input.right == UP && (!duck || physic.get_velocity_y() != 0)) {
      old_dir = dir;
      dir = LEFT;
      dirsign = -1;
  } else if(input.left == UP && input.right == DOWN && (!duck || physic.get_velocity_y() != 0)) {
      old_dir = dir;
      dir = RIGHT;
      dirsign = 1;
  }

  if (input.fire == UP) {
      ax = dirsign * WALK_ACCELERATION_X;
      // limit speed
      if(vx >= MAX_WALK_XM && dirsign > 0) {
        vx = MAX_WALK_XM;
        ax = 0;
      } else if(vx <= -MAX_WALK_XM && dirsign < 0) {
        vx = -MAX_WALK_XM;
        ax = 0;
      }
  } else {
      ax = dirsign * RUN_ACCELERATION_X;
      // limit speed
      if(vx >= MAX_RUN_XM && dirsign > 0) {
        vx = MAX_RUN_XM;
        ax = 0;
      } else if(vx <= -MAX_RUN_XM && dirsign < 0) {
        vx = -MAX_RUN_XM;
        ax = 0;
      }
  }

  // we can reach WALK_SPEED without any acceleration
  if(dirsign != 0 && fabs(vx) < WALK_SPEED) {
    vx = dirsign * WALK_SPEED;
  }

  // changing directions?
  if(on_ground() && ((vx < 0 && dirsign >0) || (vx>0 && dirsign<0))) {
      if(fabs(vx)>SKID_XM && !skidding_timer.check()) {
          skidding_timer.start(SKID_TIME);
#ifndef NOSOUND
#ifndef USEMIKMOD
          play_sound(sounds[SND_SKID], SOUND_CENTER_SPEAKER);
#else
	  play_chunk(SND_SKID);
#endif
#endif
          ax *= 2.5;
      } else {
          ax *= 2;
      }
  }

  // we get slower when not pressing any keys
  if(dirsign == 0) {
      if(fabs(vx) < WALK_SPEED) {
          vx = 0;
          ax = 0;
      } else if(vx < 0) {
          ax = WALK_ACCELERATION_X * 1.5;
      } else {
          ax = WALK_ACCELERATION_X * -1.5;
      }
  }

  // if we're on ice slow down acceleration or deceleration
  if (isice(base.x, base.y + base.height))
  {
    /* the acceleration/deceleration rate on ice is inversely proportional to
     * the current velocity.
     */

    // increasing 1 will increase acceleration/deceleration rate
    // decreasing 1 will decrease acceleration/deceleration rate
    //  must stay above zero, though
    if (ax != 0) ax *= 1 / fabs(vx);
  }

  physic.set_velocity(vx, vy);
  physic.set_acceleration(ax, ay);
}

void
Player::handle_vertical_input()
{
  // Press jump key
  if(input.up == DOWN && can_jump)
    {
      if (on_ground())
        {
          // jump higher if we are running
          if (fabs(physic.get_velocity_x()) > MAX_WALK_XM)
            physic.set_velocity_y(5.8);
          else
            physic.set_velocity_y(5.2);

          --base.y;
          jumping = true;
          can_jump = false;
          if (size == SMALL) {
#ifndef NOSOUND
#ifndef USEMIKMOD
            play_sound(sounds[SND_JUMP], SOUND_CENTER_SPEAKER);
#else
	    play_chunk(SND_JUMP);
#endif
		}
          else {
#ifndef USEMIKMOD
            play_sound(sounds[SND_BIGJUMP], SOUND_CENTER_SPEAKER);
#else
	    play_chunk(SND_BIGJUMP);
#endif
#endif
          }
        }
    }
  // Let go of jump key
  else if(input.up == UP && jumping)
    {
      jumping = false;
      if(physic.get_velocity_y() > 0) {
        physic.set_velocity_y(0);
      }
    }

  if ( (issolid(base.x + base.width / 2, base.y + base.height + 64) ||
        issolid(base.x + 1, base.y + base.height + 64) ||
        issolid(base.x + base.width - 1, base.y + base.height + 64))
       && jumping  == false
       && can_jump == false
       && input.up == DOWN
       && input.old_up == UP)
    {
      can_jump = true;
    }

  input.old_up = input.up;
}

void
Player::handle_input()
{
  /* Handle horizontal movement: */
    handle_horizontal_input();

  /* Jump/jumping? */

  if (on_ground() && input.up == UP)
    can_jump = true;
  if (input.up == DOWN || (input.up == UP && jumping))
    {
      handle_vertical_input();
    }

  /* Shoot! */

  if (input.fire == DOWN && input.old_fire == UP && got_coffee)
    {
      World::current()->add_bullet(base.x, base.y, physic.get_velocity_x(), dir);
      input.old_fire = DOWN;
    }

  /* tux animations: */
  if(!frame_timer.check())
    {
      frame_timer.start(25);
      if (input.right == UP && input.left == UP)
        {
          frame_main = 1;
          frame_ = 1;
        }
      else
        {
          if ((input.fire == DOWN && (global_frame_counter % 2) == 0) ||
              (global_frame_counter % 4) == 0)
            frame_main = (frame_main + 1) % 4;

          frame_ = frame_main;

          if (frame_ == 3)
            frame_ = 1;
        }
    }

  /* Duck! */
  if (input.down == DOWN && size == BIG && !duck && physic.get_velocity_y() == 0 && on_ground())
    {
      duck = true;
      base.height = 32;                             
      base.y += 32;
      // changing base size confuses collision otherwise
      old_base = previous_base = base;
    }
  else if(input.down == UP && size == BIG && duck && physic.get_velocity_y() == 0 && on_ground())
    {
      duck = false;
      base.y -= 32;
      base.height = 64;
      // changing base size confuses collision otherwise
      old_base = previous_base = base;                        
    }
}

void
Player::grow()
{
  if(size == BIG)
    return;
  
  size = BIG;
  base.height = 64;
  base.y -= 32;

  old_base = previous_base = base;
}

void
Player::jump_of_badguy(BadGuy* badguy)
{
  if(input.up)
    physic.set_velocity_y(5.2);
  else
    physic.set_velocity_y(2.0);
  base.y = badguy->base.y - base.height-2;
}

void
Player::grabdistros()
{
  /* Grab distros: */
  if (!dying)
    {
      World::current()->trygrabdistro(base.x, base.y, NO_BOUNCE);
      World::current()->trygrabdistro(base.x+ 31, base.y, NO_BOUNCE);

      World::current()->trygrabdistro(base.x, base.y + base.height, NO_BOUNCE);
      World::current()->trygrabdistro(base.x+ 31, base.y + base.height, NO_BOUNCE);

      if(size == BIG)
        {
          World::current()->trygrabdistro(base.x, base.y + base.height / 2, NO_BOUNCE);
          World::current()->trygrabdistro(base.x+ 31, base.y + base.height / 2, NO_BOUNCE);
        }

    }

  /* Enough distros for a One-up? */
  if (player_status.distros >= DISTROS_LIFEUP)
    {
      player_status.distros = player_status.distros - DISTROS_LIFEUP;
      if(player_status.lives < MAX_LIVES)
        ++player_status.lives;
      /*We want to hear the sound even, if MAX_LIVES is reached*/
#ifndef NOSOUND
#ifndef USEMIKMOD
      play_sound(sounds[SND_LIFEUP], SOUND_CENTER_SPEAKER);
#else
      play_chunk(SND_LIFEUP);
#endif
#endif
    }
}

void
Player::draw()
{
  if (!safe_timer.started() || (global_frame_counter % 2) == 0)
    {
      if (dying == DYING_SQUISHED)
        {
          smalltux_gameover->draw(base.x - scroll_x, base.y);
        }
      else
        {
          PlayerSprite* sprite;
          
          if (size == SMALL)
            sprite = &smalltux;
          else if (got_coffee)
            sprite = &firetux;
          else
            sprite = &largetux;
          
          if (duck && size != SMALL)
            {
              if (dir == RIGHT)
                sprite->duck_right->draw(base.x - scroll_x, base.y);
              else 
                sprite->duck_left->draw(base.x - scroll_x, base.y);
            }
          else if (skidding_timer.started())
            {
              if (dir == RIGHT)
                sprite->skid_right->draw(base.x - scroll_x, base.y);
              else
                sprite->skid_left->draw(base.x - scroll_x, base.y); 
            }
          else if (kick_timer.started())
            {
              if (dir == RIGHT)
                sprite->kick_right->draw(base.x - scroll_x, base.y);
              else
                sprite->kick_left->draw(base.x - scroll_x, base.y); 
            }
          else if (physic.get_velocity_y() != 0)
            {
              if (dir == RIGHT)
                sprite->jump_right->draw(base.x - scroll_x, base.y);
              else
                sprite->jump_left->draw(base.x - scroll_x, base.y);                   
            }
          else
            {
              if (fabsf(physic.get_velocity_x()) < 1.0f) // standing
                {
                  if (dir == RIGHT)
                    sprite->stand_right->draw( base.x - scroll_x, base.y);
                  else
                    sprite->stand_left->draw( base.x - scroll_x, base.y);
                }
              else // moving
                {
                  if (dir == RIGHT)
                    sprite->walk_right->draw(base.x - scroll_x, base.y);
                  else
                    sprite->walk_left->draw(base.x - scroll_x, base.y);
                }
            }
                      
          // Draw arm overlay graphics when Tux is holding something
          if (holding_something && physic.get_velocity_y() == 0 && !duck)
            {
              if (dir == RIGHT)
                sprite->grab_right->draw(base.x - scroll_x, base.y);
              else
                sprite->grab_left->draw(base.x - scroll_x, base.y);
            }

          // Draw blinking star overlay
          if (invincible_timer.started() &&
             (invincible_timer.get_left() > TUX_INVINCIBLE_TIME_WARNING || global_frame_counter % 3))
            {
              if (size == SMALL || duck)
                smalltux_star->draw(base.x - scroll_x, base.y);
              else
                largetux_star->draw(base.x - scroll_x, base.y);
            }
        }
    }     
  
  if (debug_mode)
    fillrect(base.x - scroll_x, base.y, 
             base.width, base.height, 75,75,75, 150);
}

void
Player::collision(void* p_c_object, int c_object)
{
  BadGuy* pbad_c = NULL;

  switch (c_object)
    {
    case CO_BADGUY:
      pbad_c = (BadGuy*) p_c_object;

     /* Hurt player if he touches a badguy */
      if (!pbad_c->dying && !dying &&
          !safe_timer.started() &&
          pbad_c->mode != BadGuy::HELD)
        {
          if (pbad_c->mode == BadGuy::FLAT && input.fire == DOWN
               && !holding_something)
            {
              holding_something = true;
              pbad_c->mode = BadGuy::HELD;
              pbad_c->base.y-=8;
            }
          else if (pbad_c->mode == BadGuy::FLAT)
            {
              // Don't get hurt if we're kicking a flat badguy!
            }
          else if (pbad_c->mode == BadGuy::KICK)
            {
              /* Hurt if you get hit by kicked laptop: */
              if (!invincible_timer.started())
                {
                  kill(SHRINK);
                }
              else
                {
                   pbad_c->dying = DYING_FALLING;
#ifndef NOSOUND
#ifndef USEMIKMOD
                   play_sound(sounds[SND_FALL], SOUND_CENTER_SPEAKER);
#else
		   play_chunk(SND_FALL);
#endif
#endif
                   World::current()->add_score(pbad_c->base.x - scroll_x,
                                               pbad_c->base.y,
                                               25 * player_status.score_multiplier);
                }
            }
          else
            {
              if (!invincible_timer.started())
                {
                  kill(SHRINK);
                }
              else
                {
                  pbad_c->kill_me(25);
                }
            }
          player_status.score_multiplier++;
        }
      break;
    default:
      break;
    }

}

/* Kill Player! */

void
Player::kill(HurtMode mode)
{
#ifndef NOSOUND
#ifndef USEMIKMOD
  play_sound(sounds[SND_HURT], SOUND_CENTER_SPEAKER);
#else
  play_chunk(SND_HURT);
  updateSound();
#endif
#endif

  physic.set_velocity_x(0);

  if (mode == SHRINK && size == BIG)
    {
      if (got_coffee)
        {
          got_coffee = false;
        }
      else
        {
          size = SMALL;
          base.height = 32;
          duck = false;
        }
      safe_timer.start(TUX_SAFE_TIME);
    }
  else
    {
      physic.enable_gravity(true);
      physic.set_acceleration(0, 0);
      physic.set_velocity(0, 7);
      if(dying != DYING_SQUISHED)
      --player_status.lives;
      dying = DYING_SQUISHED;
    }
}

void
Player::is_dying()
{
  remove_powerups();
  dying = DYING_NOT;
}

bool Player::is_dead()
{
#ifndef RES320X240
  if(base.y > screen->h || base.x < scroll_x - AUTOSCROLL_DEAD_INTERVAL)  // last condition can happen in auto-scrolling
#else
  if(base.y > 640 || base.x < scroll_x - AUTOSCROLL_DEAD_INTERVAL)  // last condition can happen in auto-scrolling
#endif
    return true;
  else
    return false;
}

/* Remove Tux's power ups */
void
Player::remove_powerups()
{
  got_coffee = false;
  size = SMALL;
  base.height = 32;
}

void
Player::check_bounds(bool back_scrolling, bool hor_autoscroll)
{
  /* Keep tux in bounds: */
  if (base.x < 0)
    { // Lock Tux to the size of the level, so that he doesn't fall of
      // on the left side
      base.x = 0;
    }

  /* Keep in-bounds, vertically: */
#ifndef RES320X240
  if (base.y > screen->h)
#else
  if (base.y > 640)
#endif
    {
      kill(KILL);
#ifndef NOSOUND
#ifdef USEMIKMOD    
      float wait=SDL_GetTicks()+800;
      while ( wait > SDL_GetTicks()) {
         updateSound();
      }
#endif
#endif

    }

  if(base.x < scroll_x && (!back_scrolling || hor_autoscroll))  // can happen if back scrolling is disabled
    base.x = scroll_x;

  if(hor_autoscroll)
    {
    if(base.x == scroll_x)
      if((issolid(base.x+32, base.y) || (size != SMALL && !duck && issolid(base.x+32, base.y+32))) && (dying == DYING_NOT))
        kill(KILL);

    if(base.x + base.width > scroll_x + screen->w)
      base.x = scroll_x + screen->w - base.width;
    }
    
}

// EOF //

